File: extras/2020-go/04.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Passport map[string]string

func (p Passport) HasFields() bool {
	keys := []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"}
	for _, key := range keys {
		if _, ok := p[key]; !ok {
			return false
		}
	}
	return true
}

func (p Passport) Valid() bool {
	if !p.HasFields() {
		return false
	}

	byr, err := strconv.Atoi(p["byr"])
	assert(err)
	if byr < 1920 || byr > 2002 {
		return false
	}

	iyr, err := strconv.Atoi(p["iyr"])
	assert(err)
	if iyr < 2010 || iyr > 2020 {
		return false
	}

	eyr, err := strconv.Atoi(p["eyr"])
	assert(err)
	if eyr < 2020 || eyr > 2030 {
		return false
	}

	hgt := p["hgt"]
	isCm := strings.HasSuffix(hgt, "cm")
	isIn := strings.HasSuffix(hgt, "in")
	if !isIn && !isCm {
		return false
	}
	hgtNum, err := strconv.Atoi(hgt[:len(hgt)-2])
	assert(err)
	if isCm && (hgtNum < 150 || hgtNum > 193) {
		return false
	} else if isIn && (hgtNum < 59 || hgtNum > 76) {
		return false
	}

	hcl := p["hcl"]
	if hcl[0] != '#' {
		return false
	}
	for i := 1; i < 7; i += 1 {
		c := hcl[i]
		isNum := c >= '0' && c <= '9'
		isAl := c >= 'a' && c <= 'f'
		if !isNum && !isAl {
			return false
		}
	}

	validColors := []string{"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}
	found := false
	for _, v := range validColors {
		if v == p["ecl"] {
			found = true
			break
		}
	}
	if !found {
		return false
	}

	_, err = strconv.Atoi(p["pid"])
	if err != nil {
		return false
	}
	if len(p["pid"]) != 9 {
		return false
	}
	return true
}

func day04(input string) {
	passports := make([]Passport, 0)
	passport := make(Passport)
	for _, line := range strings.Split(input, "\n") {
		if line == "" {
			passports = append(passports, passport)
			passport = make(Passport)
			continue
		}
		for _, pair := range strings.Split(line, " ") {
			vals := strings.Split(pair, ":")
			passport[vals[0]] = vals[1]
		}
	}
	passports = append(passports, passport)
	day04a(passports)
	day04b(passports)
}

func day04a(input []Passport) {
	sum := 0
	for _, passport := range input {
		if passport.HasFields() {
			sum += 1
		}
	}
	fmt.Println("part 1:", sum)
}

func day04b(input []Passport) {
	sum := 0
	for _, passport := range input {
		if passport.Valid() {
			sum += 1
		}
	}
	fmt.Println("part 2:", sum)
}

File: 2020/04.py
import collections, itertools, math, string, util
from dataclasses import dataclass
from functools import cache
from sys import argv

F = [l.strip() for l in open(argv[1])]

S = 0
ct = 0
for l in F:
    for w in l.split():
        a,b=util.scan("%s:%s", w)
        if a == "byr" or a == "iyr" or a == "eyr" or a == "hgt" or a == "hcl" or a == "ecl" or a == "pid":
            ct+=1
    if l == "":
        if ct == 7:
            S += 1
        ct = 0
if ct == 7:
    S += 1
ct = 0
print(S)

S = 0
ct = 0
for l in F:
    for w in l.split():
        a,b=util.scan("%s:%s", w)
        if a == "byr":
            if int(b) >= 1920 and int(b) <= 2002:
                ct+=1
        if a == "iyr":
            if int(b) >= 2010 and int(b) <= 2020:
                ct+=1
        if a == "eyr":
            if int(b) >= 2020 and int(b) <= 2030:
                ct += 1
        if a=="hgt":
            l = b[:-2]
            d = b[-2:]
            if d == "cm":
                if int(l) >= 150 and int(l) <= 193:
                    ct += 1
            if d == "in":
                if int(l) >= 59 and int(l) <= 76:
                    ct += 1
        if a == "hcl":
            if b[0] == "#":
                bad = False
                for i in range(6):
                    if b[i+1] not in "abcdef0123456789":
                        bad = True
                        break
                if not bad:
                    ct += 1
        if a == "ecl":
            if b in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]:
                ct += 1
        if a == "pid":
            if len(b) == 9 and b.isdigit():
                ct += 1
    if l == "":
        if ct == 7:
            S += 1
        ct = 0
if ct == 7:
    S += 1
ct = 0
print(S)
